File: 프로그래머스/2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_map>

using namespace std;
vector<char> v={{'A','E','I','O','U'}};
vector<string> ret;
unordered_map<string,int> m;

void DFS_recur(int curr,string temp) {
    ret.push_back(temp);
    
    if(curr==5) {
        return;
    }
    
    for(int i=0;i<5;i++) {
        DFS_recur(curr+1,temp+v[i]);
    }
}
    
int solution(string word) {
    
    DFS_recur(0,"");
    sort(ret.begin(),ret.end());
    for(int i=1;i<ret.size();i++) {
        m[ret[i]]=i;
    }
    
    return m[word];
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    
    int cnt;
    char temp[]={'A','E','I','O','U'};
    HashMap<String,Integer> map = new HashMap<>();
    
    void DFS_recur(int curr,String str) {
        
        map.put(str,cnt++);
        
        if(curr==5) {
            return;
        }
        
        for(int i=0;i<5;i++) {
            DFS_recur(curr+1,str+temp[i]);
        }
    }
    public int solution(String word) {
        int answer = 0;
        
        DFS_recur(0,"");
        
        return map.get(word);
    }
}